### PR TITLE
Add events resource and event enum

### PR DESCRIPTION
**Pull request by Xaeroxe**
Fixes https://github.com/amethyst/amethyst/issues/229

Note: I opened https://github.com/amethyst/amethyst/pull/354 before but that was when my branch was malformed.  This one is correct.

### CODE REVIEW

File: src/event.rs [hunk 1]
```diff
@@ -1,4 +1,18 @@
...
-pub use winit::{ElementState, Event, KeyboardInput, MouseButton, VirtualKeyCode, WindowEvent};
+pub use winit::{ElementState, Event as WinitEvent, KeyboardInput, MouseButton, VirtualKeyCode,
+                WindowEvent};
+
+/// Events that the engine might send to the user.
+pub enum Event {
+    /// An event from the Winit crate
+    WinitEvent(WinitEvent),
+    /// Unused event, enables _ in match statements
+    UnusedEvent,
```
Comment (torkleyy) [inline src/event.rs]: I don't get why we need this variant.

File: src/event.rs [hunk 2]
```diff
@@ -1,4 +1,18 @@
...
 //! types.
 
-pub use winit::{ElementState, Event, KeyboardInput, MouseButton, VirtualKeyCode, WindowEvent};
+pub use winit::{ElementState, Event as WinitEvent, KeyboardInput, MouseButton, VirtualKeyCode,
+                WindowEvent};
+
+/// Events that the engine might send to the user.
+pub enum Event {
+    /// An event from the Winit crate
+    WinitEvent(WinitEvent),
```
Comment (torkleyy) [inline src/event.rs]: Should we make this implementation detail part of the public API?

File: src/event.rs (hunk 1 above)
Comment (omni-viral) [reply src/event.rs]: As I understand it there is only one variant for now. And @Xaeroxe use `_` arm in matches so those matches will stay valid when new variants will be added here.

Comment (omni-viral) [reply src/event.rs]: Although I was thinking `_` arm could be added even if match have arm for all variants.

File: src/event.rs (hunk 2 above)
Comment (Xaeroxe) [reply src/event.rs]: We already have actually.  Our events thus far have come from Winit.  Now the next logical question is should we leave it that way?  I think so personally, as reimplementing and converting the events could be a pain and cost us cycles unnecessarily.  Winit is also a cross platform library that should work for every use case we have thus far.

File: src/event.rs (hunk 1 above)
Comment (omni-viral) [reply src/event.rs]: Well. I've checked. It is possible but causes warning.

File: src/event.rs (hunk 2 above)
Comment (torkleyy) [reply src/event.rs]: I still see some issues. E.g. the hierarchy is deeper.

File: src/event.rs (hunk 1 above)
Comment (torkleyy) [reply src/event.rs]: Well we should make this enum future proof and add a `__NonExhaustive` variant instead.

File: src/event.rs (hunk 2 above)
Comment (Xaeroxe) [reply src/event.rs]: Yeah the hierarchy is getting kind of excessive at this point.  Perhaps I can see what I can do about that.

File: src/event.rs (hunk 1 above)
Comment (Xaeroxe) [reply src/event.rs]: Huh, I actually didn't know about `__NonExhaustive` thanks!

Comment (stenverbois) [reply src/event.rs]: It does not matter what it is called as it is not a standard feature. `__Nonexhaustive` is a std-only hack. The best we can do at the moment is a `#[doc(hidden)]` variant.

[#[non_exhaustive] is on its way](https://github.com/rust-lang/rust/issues/44109)

Comment (torkleyy) [reply src/event.rs]: @stenvorbis Yeah I don't really care how it's called, but prefixing with `__` makes sense and this name is somehow a known pattern. Ye, we should have `#[doc(hidden)]`

File: src/app.rs [hunk 3]
```diff
@@ -163,10 +163,18 @@ impl<'a, 'b> Application<'a, 'b> {
                 )
                 {
                     let mut input = engine.world.write_resource::<InputHandler>();
-                    if let Event::WindowEvent { ref event, .. } = event {
-                        input.send_event(&event);
+                    if let WinitEvent::WindowEvent { ref event, .. } = event {
+                        input.send_event(event);
                     }
                 }
+                let event = match event {
```
Comment (torkleyy) [inline src/app.rs]: I'd actually make the hierarchy even flatter, but it's definitely an improvement.

File: src/event.rs [hunk 4]
```diff
@@ -1,4 +1,23 @@
...
+#[derive(Clone)]
+pub enum Event {
+    /// A WindowEvent from the winit crate
+    WindowEvent(WindowEvent),
+    /// A DeviceEvent from the winit crate
+    DeviceEvent(DeviceId, DeviceEvent),
+    /// Awakened event from the winit crate
+    Awakened,
+    // Add more events here as needed.
+    #[doc(hidden)]
```
Comment (torkleyy) [inline src/event.rs]: Could we please have docs before attributes? I think that's most used order in this crate and also the one I prefer.

File: src/event.rs [hunk 5]
```diff
@@ -1,4 +1,23 @@
-//! This module contains the `WindowEvent` type and re-exports glutin event
+//! This module contains the `Event` enum and re-exports glutin event
 //! types.
 
-pub use winit::{ElementState, Event, KeyboardInput, MouseButton, VirtualKeyCode, WindowEvent};
+pub use winit::{DeviceEvent, DeviceId, ElementState, KeyboardInput, MouseButton, VirtualKeyCode, WindowEvent};
+
+/// Events that the engine might send to the user.
```
Comment (LucioFranco) [inline src/event.rs]: Do we want to either give a bit more detail on what is within `WindowEvent` or link to the winit crate? Might make it easier for newer people to figure out how to use the input system.

Comment (Xaeroxe) [reply src/event.rs]: Input system is managed automatically by the amethyst crate, so users aren't required to pump events themselves.  However as @torkleyy said I think I'm going to flatten this hierarchy anyways.

**Comment by Xaeroxe** [comment]
This is quite inferior to an approach using https://github.com/Rhuagh/shrev-rs so I'm going to close this and rewrite it with that later.

**Pull request closed without merging**